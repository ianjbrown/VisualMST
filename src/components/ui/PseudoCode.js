import { ListGroup } from "react-bootstrap";

function PseudoCode(props) {
  if (props.algName === "Kruskal's Algorithm") {
    return (
      <ListGroup variant="flush" as="ul">
        {props.algorithmState === 0 ? (
          <ListGroup.Item variant="warning" as="li">
            Sort edges in non-decreasing order;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            Sort edges in non-decreasing order;
          </ListGroup.Item>
        )}
        {props.algorithmState === 1 ? (
          <ListGroup.Item variant="warning" as="li">
            While edges remain in queue
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            While edges remain in queue
          </ListGroup.Item>
        )}
        {props.algorithmState === 2 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; e = edge with smallest weight;&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; e = edge with smallest weight;&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 3 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; If adding e to MST doesn't create cycle&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; If adding e to MST doesn't create cycle&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 4 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp;&emsp; Add e to MST;&emsp;&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp;&emsp; Add e to MST;&emsp;&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 5 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; Else&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; Else&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 6 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp;&emsp; Reject e from MST;&emsp;&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp;&emsp; Reject e from MST;&emsp;&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 7 ? (
          <ListGroup.Item variant="warning" as="li">
            We have found a minimum spanning tree with n-1 edges
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            We have found a minimum spanning tree with n-1 edges
          </ListGroup.Item>
        )}
      </ListGroup>
    );
  } else if (props.algName === "Prim's Algorithm") {
    return (
      <ListGroup variant="flush" as="ul">
        {props.algorithmState === 0 ? (
          <ListGroup.Item variant="warning" as="li">
            Initialise visited set with starting vertex
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            Initialise visited set with starting vertex
          </ListGroup.Item>
        )}
        {props.algorithmState === 1 ? (
          <ListGroup.Item variant="warning" as="li">
            Add connecting edges to queue.
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            Add connecting edges to queue.
          </ListGroup.Item>
        )}
        {props.algorithmState === 2 ? (
          <ListGroup.Item variant="warning" as="li">
            While visited doesn't contain all vertices
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            While visited doesn't contain all vertices
          </ListGroup.Item>
        )}
        {props.algorithmState === 3 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; select edge with smallest weight&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; select edge with smallest weight&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 4 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; If corresponding vertex not visited&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; If corresponding vertex not visited&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 5 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp;&emsp; Add vertex to visited and adjacent edges to queue
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp;&emsp; Add vertex to visited and adjacent edges to queue
          </ListGroup.Item>
        )}
        {props.algorithmState === 6 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp; Else&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp; Else&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 7 ? (
          <ListGroup.Item variant="warning" as="li">
            &emsp;&emsp;Ignore edge&emsp;&emsp;
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            &emsp;&emsp; Ignore edge&emsp;&emsp;
          </ListGroup.Item>
        )}
        {props.algorithmState === 8 ? (
          <ListGroup.Item variant="warning" as="li">
            Found minimum spanning tree
          </ListGroup.Item>
        ) : (
          <ListGroup.Item variant="dark" as="li">
            Found minimum spanning tree
          </ListGroup.Item>
        )}
      </ListGroup>
    );
  }
}

export default PseudoCode;
