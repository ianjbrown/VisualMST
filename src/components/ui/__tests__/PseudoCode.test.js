import { render, screen, cleanup } from "@testing-library/react";
import PseudoCode from "../PseudoCode";
import "@testing-library/jest-dom/";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});

const kPseudoLines = [
  "Sort edges in non-decreasing order;",
  "While edges remain in queue",
  "e = edge with smallest weight;",
  "If adding e to MST doesn't create cycle",
  "Add e to MST;",
  "Else",
  "Reject e from MST;",
  "We have found a minimum spanning tree with n-1 edges",
];

const pPseudoLines = [
  "Initialise visited set with starting vertex",
  "Add connecting edges to queue.",
  "While visited doesn't contain all vertices",
  "select edge with smallest weight",
  "If corresponding vertex not visited",
  "Add vertex to visited and adjacent edges to queue",
  "Else",
  "Ignore edge",
  "Found minimum spanning tree",
];

test("Kruskal's PseudoCode matches snapshot", () => {
  const tree = renderer
    .create(<PseudoCode algName="Kruskal's Algorithm" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Prim's PseudoCode matches snapshot", () => {
  const tree = renderer
    .create(<PseudoCode algName="Prim's Algorithm" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Kruskal's PseudoCode loads and has correct attributes", () => {
  render(<PseudoCode algName="Kruskal's Algorithm" />);
  const kruskalListGroup = screen.getByTestId("kruskal-list-group");
  expect(kruskalListGroup).toBeInTheDocument();
  expect(kruskalListGroup).toHaveClass("list-group list-group-flush");
});

test("Prim's PseudoCode loads and has correct attributes", () => {
  render(<PseudoCode algName="Prim's Algorithm" />);
  const primListGroup = screen.getByTestId("prim-list-group");
  expect(primListGroup).toBeInTheDocument();
  expect(primListGroup).toHaveClass("list-group list-group-flush");
});

// test kruskal pseudocode highlight for each algorithmstate.
for (let i = 0; i < kPseudoLines.length; i++) {
  test(`Kruskal's PseudoCode Highlighting ${i}`, () => {
    render(<PseudoCode algName="Kruskal's Algorithm" algorithmState={i} />);
    const highlightLine = screen.queryByText(kPseudoLines[i]);
    expect(highlightLine).toBeInTheDocument();
    expect(highlightLine).toHaveClass(
      "list-group-item list-group-item-warning"
    );

    for (let j = 0; j < kPseudoLines.length; j++) {
      if (j === i) continue;
      const line = screen.getByText(kPseudoLines[j]);
      expect(line).toBeInTheDocument();
      expect(line).toHaveClass("list-group-item list-group-item-dark");
    }
  });
}

// test prim pseudocode highlight for each algorithmstate
for (let i = 0; i < pPseudoLines.length; i++) {
  test(`Prim's PseudoCode Highlighting ${i}`, () => {
    render(<PseudoCode algName="Prim's Algorithm" algorithmState={i} />);
    const highlightLine = screen.queryByText(pPseudoLines[i]);
    expect(highlightLine).toBeInTheDocument();
    expect(highlightLine).toHaveClass(
      "list-group-item list-group-item-warning"
    );

    for (let j = 0; j < pPseudoLines.length; j++) {
      if (j === i) continue;
      const line = screen.getByText(pPseudoLines[j]);
      expect(line).toBeInTheDocument();
      expect(line).toHaveClass("list-group-item list-group-item-dark");
    }
  });
}
