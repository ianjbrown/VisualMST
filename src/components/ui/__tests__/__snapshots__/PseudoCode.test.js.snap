// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kruskal's PseudoCode matches snapshot 1`] = `
<ul
  className="list-group list-group-flush"
  data-testid="kruskal-list-group"
  onKeyDown={[Function]}
>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    Sort edges in non-decreasing order;
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    While edges remain in queue
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      e = edge with smallest weight; 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      If adding e to MST doesn't create cycle 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
       Add e to MST;  
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      Else 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
       Reject e from MST;  
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    We have found a minimum spanning tree with n-1 edges
  </li>
</ul>
`;

exports[`Prim's PseudoCode matches snapshot 1`] = `
<ul
  className="list-group list-group-flush"
  data-testid="prim-list-group"
  onKeyDown={[Function]}
>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    Initialise visited set with starting vertex
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    Add connecting edges to queue.
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    While visited doesn't contain all vertices
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      select edge with smallest weight 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      If corresponding vertex not visited 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
       Add vertex to visited and adjacent edges to queue
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
      Else 
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
       Ignore edge  
  </li>
  <li
    className="list-group-item list-group-item-dark"
    data-rr-ui-event-key={null}
    onClick={[Function]}
  >
    Found minimum spanning tree
  </li>
</ul>
`;
